### PR TITLE
ARROW-10818: [Rust] Initial implementation for Decimal 128/256

### PR DESCRIPTION
**Pull request by ovr**
Implementation for DataType::Decimal which is represented by Decimal128Type on top of i128.

This PR Implements:

- Math (Add/Sub)
- Casts (integers/floats/utf8 to Decimal and back)
- Reading CSV
- GROUP BY

Notice:

- This PR doesn't change ArrayDecimal to be generic, it's planned to be done in another PR with Decimal256 will be done.
- This PR doesn't enable decimal in DataFushion, it continues to use Float64 for Decimal

### CODE REVIEW

File: rust/arrow/src/datatypes/mod.rs [hunk 1]
```diff
@@ -199,6 +207,81 @@ pub struct Field {
...
+
+    fn from_bytes_with_precision_scale(
+        bytes: &[u8],
+        precision: usize,
+        scale: usize,
+    ) -> Self;
+}
+
+#[derive(Debug)]
+pub enum DecimalType {
```
Comment (ovr) [inline rust/arrow/src/datatypes/mod.rs:238]: @alamb @andygrove @Dandandan @nevi-me @jorgecarleitao

I created representations for Decimal as `Int32DecimalType`, `Int64DecimalType`, `Int128DecimalType` or `LargeDecimal`, which uses BigInt.

1. What is the best way in Rust to abstract representations of `DataType::Decimal` in Rust? Enum?
2. How should I work with `DecimalArray`?  Should make it as generic by passing representation type to it (but how should I detect type, match in everyplace?) or make it generic on top of DecimalType and get byte size from it?

Thanks

Comment (alamb) [reply rust/arrow/src/datatypes/mod.rs:238]: Hi @ovr  -- sorry for the lack of response. I am not familiar with this code so I will need some non trivial time to review and think about your question. I hope to find some time this weekend to do so.

**Comment by jorgecarleitao** [comment]
Hi @ovr , I went through what is here so far.

First of all, great stuff that you are taking this on.

Broadly speaking, this PR currently contains the following "steps":

1. make changes to the arrow crate to enable Decimal type without adding any new functionality (e.g. rename `datatypes.rs`) 
2. add the native type to the arrow crate
3. add the Array type to the arrow crate
4. add support to parquet
5. add support to DataFusion

I suggest that 4-5 are done on a separate PR. 1-3 can be done on this PR, but I suggest that we split them on separate commits.

Concerning step 2, the two main checks in the list are:

1. the type's logical representation exists in the arrows' spec
2. the type's physical representation matches arrows' spec

Looking at [the spec](https://github.com/apache/arrow/blob/master/format/Schema.fbs#L186), a decimal type only supports 128 and 256 bits. So, I am do not understand why we are trying to add support for `Int32,Int64,Int128,LargeDecimal` here. Since Rust only supports `i128`, I would say that we should only go for `i128` (the same reason we do not support `f16`).

A related aspect is that `bigint`'s official documentation states

> DEPRECATED
> This crate is deprecated and will not be developed further. Users are invited to prefer the uint crate instead.

thus, I am not sure we want to take that as a dependency.

Could you clarify some of these points?

File: rust/arrow/src/compute/kernels/cast.rs [hunk 2]
```diff
@@ -443,6 +456,13 @@ pub fn cast(array: &ArrayRef, to_type: &DataType) -> Result<ArrayRef> {
             ))),
         },
 
+        // start decimal casts
+        (Int8, Decimal(p, s)) => cast_numeric_to_decimal::<Int8Type>(array, *p, *s),
+        (Int16, Decimal(p, s)) => cast_numeric_to_decimal::<Int16Type>(array, *p, *s),
+        (Int32, Decimal(p, s)) => cast_numeric_to_decimal::<Int32Type>(array, *p, *s),
+        (Int64, Decimal(p, s)) => cast_numeric_to_decimal::<Int64Type>(array, *p, *s),
+        // end numeric casts
```
Comment (alamb) [inline rust/arrow/src/compute/kernels/cast.rs:464]: I think this means to me

```suggestion
        // end decimal casts
```

**Review by alamb — COMMENTED** [review]
First of all thank you @ovr  for this proposal and contribution. I also finally had a chance to go through this PR this morning and I broadly agree with @jorgecarleitao 's recommendations

Here are some additional thoughts:
* My reading of the [arrow spec](https://github.com/apache/arrow/blob/master/format/Schema.fbs#L186),  also suggests that Arrow only defines the `Int128Decimal` and `Int256Decimal`. In my opinion, we should not add types to the Rust implementation that are not in the arrow spec. 
* If we want to add arbitrary precision decimals to Arrow, perhaps we should propose that to the arrow mailing list and see what other implementers say
* https://github.com/rust-num/num-bigint looks like a solid choice of crate to base an arrow decimal implementation off of

> How should I work with DecimalArray? Should I make it as generic by passing representation type to it (but how should I detect type, match in everyplace? As I have known functions cannot return type) or make it generic on top of DecimalType and get byte size from it?

I am not quite what you are asking, but the common pattern for passing arrays around is as an `ArrayRef` and then downcasting to the appropriate type. 

Given the implementation in this PR that is backed by different bit widths, I see you can't really treat all DecimalArrays the same. The normal arrow pattern I would expect would something like add a `DataType::Decimal128` and  `DataType::Decimal256` type for the different underlying  bit widths and then use them as follows:
```
let decimal_array: DecimalArray128 = make_a_decimal_array();

let array : ArrayRef = Arc::new(decimal_array);
// now pass along array as most of the rest of Arrow library deals in `ArrayRef`
```

To go from `array` back to `DecimalArray128` the pattern looks like:

```
match array.data_type() {
...
  DataType::Decimal128(precision, scale) => {
    let decimal_array = array.as_any().downcast_ref::<DecimalArray128>().unwrap();
    // do whatever you need to do
 }
```

File: rust/arrow/src/compute/kernels/cast.rs (hunk 2 above)
Comment (ovr) [reply rust/arrow/src/compute/kernels/cast.rs]: Thanks

**Comment by ovr** [comment]
> Looking at the spec, a decimal type only supports 128 and 256 bits. So, I am not understanding why we are trying to add support for Int32,Int64,Int128,LargeDecimal here.

Arrow is used inside DF, which is used to build databases on to of it. If the user defines `DECIMAL(2,2)` it's a bit overhead to use i128 as representation for it.

1. Memory usage, `DecimalArray<Decimal128>` will use 16 bytes for each element.
2. Performance, Rust is using LLVM, which allows integers of any bit width from 1 to 2^23. Inside LLVM backend for x86-64, It will use rax, rcx to store it.

It's ok to store DECIMAL(2,2) as Int64 inside parquet (but anyway overhead), but it's not great to use it as memory type.

**Comment by jorgecarleitao** [comment]
> Arrow is used inside DF, which is used to build databases on to of it. If the user defines `DECIMAL(2,2)` it's a bit overhead to use i128 as representation for it.

@ovr, maybe there is some misunderstanding of the Arrow format and the scope of this crate.

Arrow format's goal is to address a common problem in performing analytics: interoperability of data between implementations. For example, the internal memory representation of data in spark is different from Python's numpy, which means that using some numpy function against that data is expensive because it incurs a complete serialization-deserialization roundtrip to handle spark datum.

The goal of this crate is to offer the necessary implementation in Rust to allow handling the arrow format in memory.

If we start adding extra types to the crate, we fall to the exact same problem that we were trying to solve in the first place: other implementations will need to implement a custom serialize and deserialize and therefore they will have to both implement it and incur the serialization-deserialization roundtrip cost.

This is the reason for @alamb 's comment that, if you think that arrow would benefit from other decimal types, then the right place to do it is on a broader forum that includes all other implementations (and the discussion of the spec itself).

Does this make sense?

> Memory usage, DecimalArray<Decimal128> will use 16 bytes for each element.

The argument is not about memory usage, but about memory layout and how it is consistent with a specification. Arrow specification is very specific about how memory is lay out so that it enables a [stable ABI](https://arrow.apache.org/docs/format/CDataInterface.html) to share data within the same process via foreign interfaces as well as across processes via the [flight protocol](https://arrow.apache.org/docs/format/Flight.html).

**Comment by alamb** [comment]
Thank you for the clarification @jorgecarleitao  -- that is a great way to summarize the rationale for my suggestions.

**Comment by alamb** [comment]
@ovr  what do you think we should do with this PR? Is it worth keeping open as a Draft or shall we close it for now?

**Comment by ovr** [comment]
> @ovr what do you think we should do with this PR? Is it worth keeping open as a Draft or shall we close it for now?

It's time to finish it. So, I've rebased PR and drop `Decimal256` from it (will send it in another PR). Planing to finish it on weekends.

File: rust/datafusion/src/physical_plan/group_scalar.rs [hunk 3]
```diff
@@ -22,10 +22,12 @@ use std::convert::{From, TryFrom};
 
 use crate::error::{DataFusionError, Result};
 use crate::scalar::ScalarValue;
+use arrow::datatypes::Decimal128Type;
 
 /// Enumeration of types that can be used in a GROUP BY expression
 #[derive(Debug, PartialEq, Eq, Hash, Clone)]
 pub(crate) enum GroupByScalar {
+    Decimal128(i128, usize, usize),
```
Comment (ovr) [inline rust/datafusion/src/physical_plan/group_scalar.rs:30]: @alamb Should I store it as (i128, usize, usize) or `Decimal128Type` (which is a structure)?

Comment (ovr) [reply rust/datafusion/src/physical_plan/group_scalar.rs]: Change it to 

```
    // Box is needed to reduce memory usage
    Decimal128(Box<Decimal128Type>),
```

Comment (alamb) [reply rust/datafusion/src/physical_plan/group_scalar.rs]: I am not sure @ovr  -- I have sort of lost the context of this PR in my head. When you are ready let me know and I can find time to review this again.

Just to be super clear, your goal is to implement the Decimal128 and Decimal256 types that are defined in the Arrow spec, right? Not an extension?

Comment (ovr) [reply rust/datafusion/src/physical_plan/group_scalar.rs]: > your goal is to implement the Decimal128 and Decimal256 types

right. In this PR only `Decimal128 `, `Decimal256 ` will be in next PR

Comment (alamb) [reply rust/datafusion/src/physical_plan/group_scalar.rs]: Awesome -- thank you 👍

**Comment by alamb** [comment]
BTW the test workspace check is  failing on master. See more details on https://issues.apache.org/jira/browse/ARROW-11896; I don't want you to waste your time hunting down a bug that is not yours.

**Comment by alamb** [comment]
FYI I merged https://github.com/apache/arrow/pull/9653 / [ARROW-11896](https://issues.apache.org/jira/browse/ARROW-11896) for the Rust CI checks which may affect this PR. If you see "Rust / AMD64 Debian 10 Rust stable test workspace" failing  with a linker error or no logs, rebasing against master will hopefully fix the problem

**Comment by jorgecarleitao** [comment]
If we want to support `Decimal(i128)`, can't we use Rust's native `i128`? E.g. does `decimal128 + decimal128` correspond to Rust's `i128 + i128`, or is the math actually different?

**Comment by ovr** [comment]
> If we want to support Decimal(i128), can't we use Rust's native i128? E.g. does decimal128 + decimal128 correspond to Rust's i128 + i128, or is the math actually different?

@jorgecarleitao 

It's using `i128` under the hood as expected (which generates to u64 and i64 by LLVM). As you know, `Decimal` has precision/scale and It's not possible to calculate the sum of two different decimals with different scales without rescaling, for example: `Decimal(1,1)` and `Decimal(1,5)`.

Yet another thing, it's overflowing.

`DECIMAL(1) + DECIMAL(1) = DECIMAL(1)`, example `1` + `9`.

**Comment by ovr** [comment]
@alamb @jorgecarleitao Can you take a look and do a review? Thanks!

**Comment by alamb** [comment]
Thanks @ovr ! I have put this on my review queue -- but I will likely not be able to do so until tomorrow

File: rust/arrow/src/array/builder.rs [hunk 4]
```diff
@@ -1214,11 +1214,12 @@ impl DecimalBuilder {
...
     }
 
-    /// Appends a byte slice into the builder.
+    /// Append i128 to into the builder. For DecimalNType please use append_value.
+    /// This method is useful when you are reading a data and doesnt require to cast value to DecimalNType
     ///
     /// Automatically calls the `append` method to delimit the slice appended in as a
     /// distinct array element.
-    pub fn append_value(&mut self, value: i128) -> Result<()> {
+    pub fn append_value_i128(&mut self, value: i128) -> Result<()> {
```
Comment (alamb) [inline rust/arrow/src/array/builder.rs:1222]: What about something like 

```suggestion
    pub fn append_value(&mut self, value: impl Into<ArrowDecimalType>) -> Result<()> {
      let value: ArrowDecimalType = value.into();
```

And then as long as you implement `impl From<i128>` for Decimal128Type` it should be possible to call `append_value` with either a `Decimal128Type` or an `i128`

File: rust/arrow/src/compute/kernels/take.rs [hunk 5]
```diff
@@ -524,6 +528,75 @@ where
...
+        nulls,
+        0,
+        vec![buffer.into()],
+        vec![],
+    );
+
+    Ok(DecimalArray::from(Arc::new(data) as ArrayDataRef))
+}
+
 /// `take` implementation for string arrays
```
Comment (alamb) [inline rust/arrow/src/compute/kernels/take.rs:600]: I suggest adding some coverage for the `take` kernels with `DecimalArrays`

File: rust/arrow/src/csv/reader.rs [hunk 6]
```diff
@@ -633,6 +636,45 @@ fn build_primitive_array<T: ArrowPrimitiveType + Parser>(
...
+
+                let parsed = match Decimal128Type::parse(s, precision, scale) {
+                    Ok(number) => number,
+                    Err(_) => {
+                        return Err(ArrowError::ParseError(format!(
+                            // TODO: we should surface the underlying error here.
+                            "Error while parsing value {} for column {} at line {}",
+                            s,
+                            col_idx,
+                            line_number + row_index
```
Comment (alamb) [inline rust/arrow/src/csv/reader.rs:664]:
```suggestion
                            "Error while parsing value {} for column {} at line {}: {}",
                            s,
                            col_idx,
                            line_number + row_index,
                            e
```

might be one way of surfacing the underlying error

File: rust/arrow/src/datatypes/decimal.rs [hunk 7]
```diff
@@ -0,0 +1,926 @@
...
+}
+
+// This types are disabled, because Arrow doesnt declare Decimals for 32 / 64
+// i32 max  -              2_147_483_647i32
+// make_type!(Decimal32Type, i32, 9);
+// i64 max  -  9_223_372_036_854_775_807i64
+//make_type!(Decimal64Type, i64, 18);
+
+// i128 max - 170_141_183_460_469_231_731_687_303_715_884_105_727i128
+make_type!(Decimal128Type, i128, 38);
```
Comment (alamb) [inline rust/arrow/src/datatypes/decimal.rs:628]: 👍  for comments explaining why this macro is as it is

File: rust/arrow/src/datatypes/decimal.rs [hunk 8]
```diff
@@ -0,0 +1,926 @@
...
+            fn to_isize(&self) -> Option<isize> {
+                unimplemented!("Unimplemented to_isize for {}", stringify!($name))
+            }
+
+            fn to_usize(&self) -> Option<usize> {
+                unimplemented!("Unimplemented to_usize for {}", stringify!($name))
+            }
+
+            fn to_i8(&self) -> Option<i8> {
+                Some(self.get_signed_lead_part() as i8)
```
Comment (alamb) [inline rust/arrow/src/datatypes/decimal.rs:314]: I feel like this is likely to `panic!` (at least in debug builds) if the lead part is greater than 128 / less than -128. Would it perhaps be better to check for overflow and return `None`?

File: rust/arrow/src/datatypes/decimal.rs [hunk 9]
```diff
@@ -0,0 +1,926 @@
...
+            fn hash_slice<H: Hasher>(_data: &[Self], _state: &mut H) where
+                Self: Sized, {
+                unimplemented!("Unimplemented hash_slice for {}", stringify!($name))
+            }
+        }
+
+        impl PartialEq<$name> for $name {
+            #[inline]
+            fn eq(&self, rhs: &$name) -> bool {
+                // @todo What is a correct behaviour for it? Rescaling?
```
Comment (alamb) [inline rust/arrow/src/datatypes/decimal.rs:578]: this seems like an important TODO...

File: rust/arrow/src/datatypes/decimal.rs [hunk 10]
```diff
@@ -0,0 +1,926 @@
...
+                false,
+            ),
+        ];
+
+        for (left, right, expected) in values {
+            assert_eq!(left < right, expected);
+        }
+    }
+
+    #[test]
```
Comment (alamb) [inline rust/arrow/src/datatypes/decimal.rs:843]: tests for `==` (aka `PartialEq`) are probably important

File: rust/arrow/src/datatypes/decimal.rs (hunk 7 above)
Comment (alamb) [inline rust/arrow/src/datatypes/decimal.rs:628]:
```suggestion
/// Type which stores a decimal number using 128 bits
make_type!(Decimal128Type, i128, 38);
```

File: rust/arrow/src/record_batch.rs [hunk 11]
```diff
@@ -132,10 +132,12 @@ impl RecordBatch {
...
             for (i, column) in columns.iter().enumerate() {
                 if column.len() != len {
-                    return Err(ArrowError::InvalidArgumentError(
-                        "all columns in a record batch must have the same length"
-                            .to_string(),
-                    ));
+                    return Err(ArrowError::InvalidArgumentError(format!(
+                        "all columns in a record batch must have the same length, expected {:?} but found {:?} at column {}",
+                        len,
+                        column.len(),
```
Comment (alamb) [inline rust/arrow/src/record_batch.rs:138]: I think the order of these arguments is not correct (data type should be second, right?)

File: rust/datafusion/src/physical_plan/group_scalar.rs [hunk 12]
```diff
@@ -120,12 +131,9 @@ mod tests {
     use crate::error::DataFusionError;
 
     macro_rules! scalar_eq_test {
-        ($TYPE:expr, $VALUE:expr) => {{
-            let scalar_value = $TYPE($VALUE);
-            let a = GroupByScalar::try_from(&scalar_value).unwrap();
-
-            let scalar_value = $TYPE($VALUE);
-            let b = GroupByScalar::try_from(&scalar_value).unwrap();
+        ($EXPR:expr) => {{
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/group_scalar.rs:134]: I don't understand this change -- it seems to now be testing that `try_from` returns the same value for the same input...

File: rust/datafusion/src/scalar.rs [hunk 13]
```diff
@@ -210,6 +213,14 @@ impl ScalarValue {
     /// Converts a scalar value into an array of `size` rows.
     pub fn to_array_of_size(&self, size: usize) -> ArrayRef {
         match self {
+            ScalarValue::Decimal128(e, p, s) => match e {
```
Comment (alamb) [inline rust/datafusion/src/scalar.rs:216]: I don't understand this limitation

File: rust/datafusion/tests/sql.rs [hunk 14]
```diff
@@ -366,6 +366,27 @@ async fn csv_query_group_by_float32() -> Result<()> {
...
+#[tokio::test]
+async fn csv_query_group_by_decimal() -> Result<()> {
+    let mut ctx = ExecutionContext::new();
+    register_aggregate_simple_csv(&mut ctx)?;
+
+    let sql =
+        "SELECT COUNT(*) as cnt, c4 FROM aggregate_simple GROUP BY c4 ORDER BY cnt DESC";
+    let actual = execute(&mut ctx, sql).await;
+
+    let expected = vec![
```
Comment (alamb) [inline rust/datafusion/tests/sql.rs:378]: 😮 👍

**Review by alamb — COMMENTED** [review]
This is an epic PR -- Great work @ovr. Thank you so much for all this.

I didn't cover this entire PR as closely as I would have liked due to its size, but I did spend a significant amount of time reviewing it and left a bunch of feedback 

Some of these comments  could be addressed as potential follow on PRs if you prefer, but I think some are worth paying attention to.

One thing I was thinking was perhaps we want to mark `DecimalArray` and `Decimal128Type` as "Beta" or subject to change so as to allow further change without confusing users that this is likely to be an area subject to potential change for awhile

File: rust/arrow/src/datatypes/decimal.rs [hunk 15]
```diff
@@ -0,0 +1,926 @@
...
+
+            fn to_f64(&self) -> Option<f64> {
+                // @todo Optimize this
+                Some(self.to_string().parse::<f64>().unwrap())
+            }
+        }
+
+        impl ToString for $name {
+            fn to_string(&self) -> String {
+                println!("<{},{}>({})", self.digits, self.precision, self.scale);
```
Comment (jorgecarleitao) [inline rust/arrow/src/datatypes/decimal.rs:366]: print.

**Review by jorgecarleitao — CHANGES_REQUESTED** [review]
Hi @ovr , thank you for this PR and for this work.

Unfortunately, what I feared would be happening. Unfortunately the crate does not have a well documented design, which adds a lot of confusion on how to implement new types.

IMO this PR breaks 2 designs of this crate, and requires some changes for me to support it.

## Issue 1

Let's first walk to the overall principle: all "getters" on the crate return the physical representation of a value without any logical content associated with it. For example, an iterator of `Timestamp(Second, None)` returns `i64`, not `TimestampType(i64, Second, None)`. This is by design; the `DataType` information is carried over by the Array's `DataType`, not by the individual values of that array, so that the physical representation of the getters equals the in-memory representation.

In this PR, `Array::value`, `iter`, etc. are not returning the value at position i; they are allocating a new struct, `Decimal128Type`, that contains logical information about its value (`scale` and `precision`). This creates a new level of indirection by converting, during operations, the physical memory `i128` (or equivalent representation) into a `struct {}` with scale and precision, which violates arrows' assumption that the in-memory representation follows a well defined convention. In other words, the struct `Decimal128Type` has a different in-memory layout than arrows' layout, which violates arrows' design.

The consequence is that we do not leverage the benefits of arrow for compute and many other things; a generic operation `op(array_a, array_b)` will allocate new `4*N^2` values: `2 u64s` per call of `value()` on each side. The same happens to the values' iterator, that allocates a new struct per `next`. This forbids the compiler from making optimizations and many other things (e.g. memory footprint).

## Issue 2

Note that the memory layout of a Decimal type is the same as of a `PrimitiveArray`; the only difference is that operations between two Decimal types require different semantics, exactly like operations between `Timestamp` (e.g. casting a Timestamp to another Timestamp requires some unit conversion; it is not simply `i64 -> i64`).

The problem is that `DecimalType` cannot be implemented on a `PrimitiveArray` because this crate has a design flaw on which `ArrowNativeType` has a constant `DataType`, but `DecimalType` has two non-enum variable values, thereby forbidding us from implementing `ArrowNativeType` for the DataType::DecimalType. An equivalent problem already exists for `Timestamp(_, str)`, where `str` is a large enum and thus we do not implement `ArrowNativeType` for each different timezone.

For this reason, we must implement another array, `DecimalArray<T: NativeType>`, and copy-paste all `PrimitiveArray<T>` stuff except the constructors, that now require a `DataType` (or a scale and precision, just like `PrimitiveArray<T>::from_opt_vec` requires a timezone).

These are the steps that IMO we would need to address issue 2:

1. declare a struct that implements `NativeType` (e.g. `[i64; 2]` for `128` and `[i64; 4]`  for `i256`, or simply implement `NativeType` for `i128`).

2. Change the constraint on `Buffer::typed_data_mut` to accept `T: NativeType` instead of `ArrowNativeType`
3. change `impl<T: ArrowNativeType> ToByteSlice` ->  `impl<T: NativeType> ToByteSlice` so that we flag to the Buffer that it can receive `NativeType`.

These are the steps that IMO we need to address issue 1:

1. Remove `Decimal128Type`
2. Make `DecimalArray::value() -> T` 
3. Make `DecimalIter<Item=Option<Option<T>>>`
4. Implement the traits that `Decimal128Type` on the new struct that implements `NativeType`

**Comment by ovr** [comment]
Thank you @alamb @jorgecarleitao for your such detailed review. Appreciate your work! I am going to rework this PR to finish it.

File: rust/arrow/src/compute/kernels/cast.rs [hunk 16]
```diff
@@ -1511,6 +1641,65 @@ mod tests {
...
+
+        assert_eq!("1.0000000000", c.value(0).to_string());
+        assert_eq!("2.0000000000", c.value(1).to_string());
+    }
+
+    #[test]
+    fn test_cast_f64_to_decimal() {
+        let a = Float64Array::from(vec![1.0, 2.0]);
+        let array = Arc::new(a) as ArrayRef;
+        let b = cast(&array, &DataType::Decimal(5, 10)).unwrap();
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/cast.rs:1659]: My understanding is that scale should be <= precision, such that we can't have `Decimal(5, 10)`. Is this correct, or is it me not understanding the rules?

File: rust/arrow/src/compute/kernels/cast.rs [hunk 17]
```diff
@@ -1511,6 +1641,65 @@ mod tests {
...
+            .unwrap();
+        builder
+            .append_value(Decimal128Type::new(654321_i128, 3, 5))
+            .unwrap();
+
+        let array = Arc::new(builder.finish()) as ArrayRef;
+        let b = cast(&array, &DataType::Float64).unwrap();
+        let c = b.as_any().downcast_ref::<Float64Array>().unwrap();
+
+        // string compare to skip clippy::float_cmp
```
Comment (nevi-me) [inline rust/arrow/src/compute/kernels/cast.rs:1698]: You can also ignore the clippy lint here if it helps :)

**Review by nevi-me — COMMENTED** [review]
Thanks for this @ovr, and I apologise for not being able to review this before.

I have a question on precision vs scale in one of the tests

**Comment by nevi-me** [comment]
Hey @ovr, do you mind separating further commits, instead of rebasing all changes into 1 commit? It makes it difficult to see what exactly has changed when you update the code.

We in any case squash everything into 1 commit. Thanks :)

File: rust/arrow/src/compute/kernels/cast.rs (hunk 16 above)
Comment (ovr) [reply rust/arrow/src/compute/kernels/cast.rs]: Not, it can be. I don't see any limitations.

**Comment by ovr** [comment]
@jorgecarleitao Can you take a look at it again? Thanks. Did I understand you correctly? Now Decimal types has ::NATIVE types, which represents a native type. `DecimalArray` started to be generic and store values as native type (I didn't change builder for it, because PR is epic large). It doesn't allocate any new structure if it doesn't need it. BTW: I created a polyfill for `i256` to test requirements for tests correctly.

**Comment by alamb** [comment]
The Apache Arrow Rust community is moving the Rust implementation into its own dedicated github repositories [arrow-rs](https://github.com/apache/arrow-rs) and [arrow-datafusion](https://github.com/apache/arrow-datafusion). It is likely we will not merge this PR into this repository

Please see the [mailing-list](https://lists.apache.org/thread.html/rce7c61cb3f703367dc00984530016e3fcb828e5a8035655fbcccf862%40%3Cdev.arrow.apache.org%3E) thread for more details

We expect the process to take a few days and will follow up with a migration plan for the in-flight PRs.

**Comment by alamb** [comment]
https://github.com/apache/arrow/pull/10096 has removed the arrow implementation from this repository (it now resides in https://github.com/apache/arrow-rs and https://github.com/apache/arrow-datafusion) in the hopes of streamlining the development process

Please re-target this PR (let us know if you need help doing so) to one/both of the new repositories. 

Thank you for understanding and helping to make arrow-rs and datafusion better

**Pull request closed without merging**